lib/export/src/Export.hs

```diff
@@ -1271,3 +1274,33 @@ makeAnnotations thy p = res
       if isNothing (List.find (== "locations-report") (theoryBuiltins thy))
         then pr
         else translateTermsReport pr
+
+-- | Pull out nots in formula
+pullnots :: LNFormula -> Either LNFormula LNFormula
+pullnots fm = 
+  let fm_partially_rewritten = pullnots' fm in
+  if notsArePulledOut fm_partially_rewritten 
+    then Right fm_partially_rewritten -- in this case, formula is fully rewritten
+    else Left fm_partially_rewritten  -- Error with partially rewritten formula
+  where
+    notsArePulledOut (Not p) = notsArePulledOut' p -- top-level not expected
+    notsArePulledOut _       =  False
+
+    notsArePulledOut' (Not _)      = False -- check that there are  no nots below top level
+    notsArePulledOut' (Conn _ p q) = notsArePulledOut' p && notsArePulledOut' q
+    notsArePulledOut' (Qua _ _ p ) = notsArePulledOut' p
+    notsArePulledOut' _            = True
+
+    pullnots' f = if f /= pullStep f then pullnots' (pullStep f) else f
```

Comment:
The idea is that `pullStep` and `pullnots'` work together to pull the negations to the front. So, I'm going to go into both of them. 
`pullStep` pulls the negations in a sub-formula "one sub-formula/level" to the front **once** (if that's possible). Essentially, it does only one step to the front wherever possible. This is why the cases where we have negations that we can pull to the front are not recursive - they just pull the negation in the sub-formula one level above and don't go deeper. If no negation can somehow be pulled to the front, then the function returns the same formula without any changes.
`pullnots'` is what pulls the negations to the front **recursively**. Originally, all should've been done by `pullStep`, i.e. the recursion should've been there. The problem was that we recurse on the formula top-down (to smaller sub-formulas), but we change the smaller sub-formulas as we pull the negations. Moreover, we cannot go back up to a bigger sub-formula, see what has changed, pull the negations further (from the level of the larger sub-formula[s] to the level of even larger sub-formulas) and then dive in at where we left. Potentially, we might need to go all the way back to the highest level to be able to pull negations or we need to "wait" for the negations three levels down to be pulled to the same level so that we can pull negations at the current one.  I didn't manage to get it to consider such cases during the top-down recursion on the formula. Thus, having everything in `pullStep` meant that some transformations where not "seen". Thus, I made it so that `pullStep` does only one step to the front and `pullnots'` makes sure we exhaustively do all steps and each time consider the results of the different steps.
`pullnots'` checks if we have pulled all negations as much as possible (`if f /= pullStep f`). In this case, calling `pullStep` doesn't pull anything - it returns the same formula again, and so does `pullnots'` (`else`-branch). If the formula was changed by the `pullStep` call, then we have pulled some negations in it one level above their respective levels. Thus, we try to pull the resulting negations in the formula one further level above their respective level (`then`-branch).

Reply:
I get it: `pullnots'` computes the fixed point. To let the code speak for itself, I suggest to:
 - rename `pullnots'` to `fixedpoint` with `pullstep` being an argument (maybe `f` if you rename the formula to `phi`). 
  - add a comment to line 1281 saying "nots pulled out by `pullStep` can enable new pull-out steps, so need to compute fixed point"